contracts/Market.sol

```diff
@@ -1610,6 +1369,89 @@ contract Market is
     return (countOfRemovalsAllocated, ids, amounts, suppliers);
   }
 
+  /**
+   * @notice Allocates removals from a specific supplier to be fulfilled.
+   * @dev This function is responsible for validating and allocating the supply from a specific supplier.
+   * @param certificateAmount The total amount of NRTs for the certificate.
+   * @param supplier The only supplier address from which to purchase carbon removals in this transaction.
+   * @param vintages A set of valid vintages from which to allocate removals.
+   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
+   * @return ids An array of the removal IDs being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   */
+  function _allocateRemovalsFromSupplierSpecificVintages(
+    uint256 certificateAmount,
+    address supplier,
+    uint256[] memory vintages
+  )
+    internal
+    returns (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts,
+      address[] memory suppliers
+    )
+  {
+    (
+      countOfRemovalsAllocated,
+      ids,
+      amounts
+    ) = _allocateSupplySingleSupplierSpecificVintages({
+      amount: certificateAmount,
+      supplier: supplier,
+      vintages: vintages
+    });
+    suppliers = new address[](countOfRemovalsAllocated).fill({value: supplier});
+    return (countOfRemovalsAllocated, ids, amounts, suppliers);
+  }
+
+  /**
+   * @notice Allocates removals to fulfill an order from the specified vintages.
+   * @dev This function is responsible for validating and allocating the supply to fulfill an order.
+   * @param purchaser The address of the purchaser.
+   * @param certificateAmount The total amount for the certificate.
+   * @param vintages A set of valid vintages from which to allocate removals.
+   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
+   * @return ids An array of the removal IDs being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   */
+  function _allocateRemovalsSpecificVintages(
+    address purchaser,
+    uint256 certificateAmount,
+    uint256[] memory vintages
+  )
+    internal
+    returns (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts,
+      address[] memory suppliers
+    )
+  {
+    uint256 availableSupply = _removal.getMarketBalance();
+    _validateSupply({
+      certificateAmount: certificateAmount,
+      availableSupply: availableSupply
+    });
```

Comment:
Any reason why this is the only `_allocateRemovals...` function that calls `validateSupply`?

Reply:
yeah it's because validate supply is just looking at whether there is enough total supply in the market to fulfill the order. that's not very useful when looking at a per-supplier or vintage-specific fulfillment since you can only fulfill from a subset of the total market supply.  in those cases the checks for insufficient supply happen as early in the process as information permits without doing a bunch of extra work at the outset to come to the same conclusion that would be arrived at during the allocation anyway